### PR TITLE
[RHOSINFRA-125] Settings sources

### PR DESCRIPTION
**Pull request by obaranov**
@yfried-redhat please review

We can merge this first and then go with the provisioner image spec change (https://github.com/rhosqeauto/InfraRed/pull/139)

### CODE REVIEW

File: cli/utils.py [hunk 1]
```diff
@@ -180,24 +180,31 @@ def normalize_file(file_path):
     return file_path
 
 
-def load_yaml(filename, search_first):
+def load_yaml(filename, *search_paths):
     """Find YAML file. search default path first.
-
     :param filename: path to file
     :param search_first: default path to search first
```
Comment (yfried-redhat) [inline cli/utils.py]: doc needs to update accordingly

**Comment by yfried-redhat** [comment]
This needs a rebase

**Comment by yfried-redhat** [comment]
retest openstack

**Comment by yfried-redhat** [comment]
retest virsh

**Comment by obaranov** [comment]
rebased

File: cli/conf.py [hunk 2]
```diff
@@ -15,6 +15,64 @@
...
+        return [os.path.join(path, app_name)
+                for path in self.get_settings_dirs()]
+
+    def validate(self):
+        """
+        Validates the configuration.
+        """
+        # Validates at least one settings dir exists
+        dirs = self.get_settings_dirs()
+        if not any([os.path.exists(path)
```
Comment (aopincar) [inline cli/conf.py]: I think 'none' will be more understandable than 'not any' here

**Comment by obaranov** [comment]
I'd suggest to rebase this only after https://github.com/rhosqeauto/InfraRed/pull/188 will be merged.

File: cli/conf.py [hunk 3]
```diff
@@ -15,6 +15,64 @@
...
+                   for path in dirs]):
```
Comment (aopincar) [inline cli/conf.py]: There's no reason for this line to be in a new line (it's not too long & it's less readable also)

File: cli/conf.py [hunk 4]
```diff
@@ -43,14 +101,8 @@ def load_config_file():
...
-    if not os.path.exists(settings_dir):
-        raise exceptions.IRFileNotFoundException(
-            settings_dir,
-            "Settings directory doesn't exist: ")
-
     return _config
 
 
-config = load_config_file()
+config = ConfigWrapper(load_config_file())
```
Comment (aopincar) [inline cli/conf.py]: What do you think about moving the 'load_config_file' function into the 'ConfigWrapper' class?

File: cli/conf.py [hunk 5]
```diff
@@ -43,14 +101,8 @@ def load_config_file():
...
+config.validate()
```
Comment (aopincar) [inline cli/conf.py]: This can be auto called if implementing the comment on the previous line.

File: cli/main.py [hunk 6]
```diff
@@ -101,27 +101,29 @@ class IRSubCommand(object):
...
         :param spec_name:  tha spec name
-        :param settings_dir:  the default spec settings dir
+        :param settings_dirs:  the default spec settings dirs
         :param args: the spec args
         :return: the IRSubCommand instance.
         """
         if args:
-            settings_dir = os.path.join(settings_dir,
-                                        spec_name)
+            settings_dirs = [os.path.join(settings_dir,
```
Comment (aopincar) [inline cli/main.py]: Please try to indent this list comprehension

File: cli/main.py [hunk 7]
```diff
@@ -101,27 +101,29 @@ class IRSubCommand(object):
...
+                                          spec_name) for settings_dir in
+                             settings_dirs]
         if "command0" in args:
-            settings_dir = os.path.join(settings_dir,
-                                        args['command0'])
-        return cls(args['command0'], args, settings_dir)
+            settings_dirs = [os.path.join(settings_dir,
```
Comment (aopincar) [inline cli/main.py]: Please try to indent list comprehension

File: tests/test_provision.py [hunk 8]
```diff
@@ -29,7 +29,7 @@ def test_dynamic_topology(tmpdir):
     # prepare config
     app_path = os.path.join(root_dir.strpath, "..")
 
-    spec.TopologyArgument.settings_dir = app_path
+    spec.TopologyArgument.settings_dirs = [app_path, ]
```
Comment (aopincar) [inline tests/test_provision.py]: comma isn't needed at the end of the list

File: tests/test_spec.py [hunk 9]
```diff
@@ -126,15 +126,15 @@ def test_yamls_file_locations(tmpdir):
     """)
 
     locations = YamlFileArgument.get_file_locations(
-        app_dir.strpath, subcommand, "arg1-arg2")
+        [app_dir.strpath, ], subcommand, "arg1-arg2")
```
Comment (aopincar) [inline tests/test_spec.py]: comma isn't needed at the end of the list

File: tests/test_spec.py [hunk 10]
```diff
@@ -126,15 +126,15 @@ def test_yamls_file_locations(tmpdir):
...
 
     assert len(locations) == 3
     assert locations[0] == file2.dirname
     assert locations[1] == file1.dirname
     assert locations[2] == "."
 
     files = YamlFileArgument.get_allowed_files(
-        app_dir.strpath, subcommand, "arg1-arg2")
+        [app_dir.strpath, ], subcommand, "arg1-arg2")
```
Comment (aopincar) [inline tests/test_spec.py]: comma isn't needed at the end of the list

**Comment by aopincar** [comment]
Comments are done from my side.
@obaranov please create a POC job that using multiple settings dirs.

**Comment by obaranov** [comment]
POC job passed: http://10.8.180.145:8080/job/poc-ir-obaranov-settings-sources/12/console

File: cli/spec.py [hunk 11]
```diff
@@ -62,11 +63,12 @@ def init_missing_args(cls, spec, args, settings_dir=None, subcommand=None):
 
         :param spec: dict. spec file
         :param args: dict. argparse arguments.
-        :param app_settings_dir: path to the base directory holding the
-            application's settings. App can be provisioner\installer\tester
+        :param settings_dirs: the list of path to the base directory
```
Comment (yfried-redhat) [inline cli/spec.py]: "A list of paths to..."

File: cli/spec.py [hunk 12]
```diff
@@ -143,32 +145,35 @@ class YamlFileArgument(ValueArgument):
...
 
     @classmethod
-    def get_file_locations(cls, settings_dir, subcommand, arg_name):
+    def get_file_locations(cls, settings_dirs, subcommand, arg_name):
         """
         Get the possible locations (folders) where the
         yaml files can be stored.
 
-        :param settings_dir: path to the base directory holding the
+        :param settings_dirs: path to the base directory holding the
```
Comment (yfried-redhat) [inline cli/spec.py]: "A list of paths to base directories..."

File: cli/spec.py [hunk 13]
```diff
@@ -143,32 +145,35 @@ class YamlFileArgument(ValueArgument):
...
-                                     *arg_name.split("-"))
-        return search_first, search_second, "."
+        search_locations = [os.path.join(
+            settings_path, subcommand, *arg_name.split("-"))
+                            for settings_path in settings_dirs]
+        root_locations = [os.path.join(
+            settings_path, *arg_name.split("-"))
+                          for settings_path in settings_dirs]
+        search_locations.extend(root_locations)
+        search_locations.append(".")
```
Comment (yfried-redhat) [inline cli/spec.py]: what's the difference between `"."` and `os.getcwd()`?

File: cli/spec.py [hunk 14]
```diff
@@ -143,32 +145,35 @@ class YamlFileArgument(ValueArgument):
...
 
     @classmethod
-    def get_allowed_files(cls, settings_dir, subcommand, arg_name,
+    def get_allowed_files(cls, settings_dirs, subcommand, arg_name,
                           search_root=False, file_extensions=('yml', 'yaml')):
         """
         Gets the list of the the files in the default locations.
 
-        :param settings_dir: path to the base directory holding the
+        :param settings_dirs: path to the base directory holding the
```
Comment (yfried-redhat) [inline cli/spec.py]: "A list of paths to the base directories..."

File: docs/source/contributors.rst [hunk 15]
```diff
@@ -23,11 +23,11 @@ InfraRed extends the ``clg`` and ``argpars`` packages with the following types
 that need to be defined in `.spec` files:
 
 * **Value**: String values
-* **YamlFile**: Expects path to YAML files. Will search for files in the settings directory before trying to resolve
   absolute path. For the argument name is "arg-name" and of subparser "SUBCOMMAND" of command "COMMAND", the default
+* **YamlFile**: Expects path to YAML files. Will search for files in one of the configured settings directories before trying to resolve absolute path. For the argument name is "arg-name" and of subparser "SUBCOMMAND" of command "COMMAND", the default
```
Comment (yfried-redhat) [inline docs/source/contributors.rst]: replace:` For the argument name` with: ` If the argument name`

File: cli/spec.py (hunk 13 above)
Comment (obaranov) [inline cli/spec.py]: no difference. changing to os usage.

File: infrared.cfg.example [hunk 16]
```diff
@@ -2,7 +2,7 @@
 # ===========================
 
 [defaults]
-settings  = settings
+settings  = settings:private_settings
```
Comment (yfried-redhat) [inline infrared.cfg.example]: cfg.example should be ready to use. private_settings isn't ready to use... 
Please remove it.

**Comment by obaranov** [comment]
POC passed: http://10.8.180.145:8080/job/poc-ir-obaranov-settings-sources/13/console

**Pull request merged**